src/hypertable_insert.c

```diff
@@ -21,6 +21,29 @@
 #include "chunk_dispatch_plan.h"
 #include "hypertable_cache.h"
 
+static PlanState *
+get_chunk_dispatch_state(ModifyTableState *mtstate, PlanState *substate)
+{
+	switch (nodeTag(substate))
+	{
+		case T_CustomScanState:
+		{
+			CustomScanState *csstate = castNode(CustomScanState, substate);
+
+			if (strcmp(csstate->methods->CustomName, CHUNK_DISPATCH_STATE_NAME) == 0)
+				return substate;
+
+			break;
+		}
+		case T_ResultState:
+			return get_chunk_dispatch_state(mtstate, castNode(ResultState, substate)->ps.lefttree);
```

Comment:
Now we only recurse into the left tree, not the right tree. Can a custom scan state be in the right tree?

Reply:
Right tree is always NULL in Result nodes: https://github.com/postgres/postgres/blob/a9cf48a4cf0c878684a2f52a3a88e29399b2065e/src/backend/optimizer/plan/createplan.c#L6745

I think it is common practice to use only the left branch in nodes that support only one child.

Reply:
Looks like the gating plan uses only `lefttree` here according to the `create_gating_plan()` and `make_result()` functions

Reply:
The corresponding place where the ResultState node is created: https://github.com/postgres/postgres/blob/a9cf48a4cf0c878684a2f52a3a88e29399b2065e/src/backend/executor/nodeResult.c#L210

(outerPlanState == left tree)